src/Akeneo/Bundle/StorageUtilsBundle/Elasticsearch/Cursor/Cursor.php

```diff
@@ -0,0 +1,152 @@
+ */
+class Cursor implements CursorInterface
+{
+    /** @var ElasticsearchClientInterface */
+    protected $esClient;
+
+    /** @var CursorableRepositoryInterface */
+    protected $repository;
+
+    /** @var array */
+    protected $esQuery;
+
+    /** @var string */
+    protected $indexType;
+
+    /** @var array */
+    protected $items = [];
+
+    /** @var int */
+    protected $pageSize;
+
+    /**
+     * @param ElasticsearchClientInterface  $esClient
+     * @param CursorableRepositoryInterface $repository
+     * @param array                         $esQuery
+     * @param string                        $indexType
+     * @param int                           $pageSize
+     */
+    public function __construct(
+        ElasticsearchClientInterface $esClient,
+        CursorableRepositoryInterface $repository,
+        array $esQuery,
+        $indexType,
+        $pageSize
+    ) {
+        $this->repository = $repository;
+        $this->esClient = $esClient;
+        $this->esQuery = $esQuery;
+        $this->indexType = $indexType;
+        $this->pageSize = $pageSize;
+
+        $this->items = $this->getNextItems($esQuery);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function current()
+    {
+        return current($this->items);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function next()
+    {
+        $currentItem = $this->current();
+
+        if (false === next($this->items)) {
+            $searchAfterIdentifier = $currentItem->getIdentifier();
+            $this->items = $this->getNextItems($this->esQuery, $searchAfterIdentifier);
+            $this->rewind();
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function key()
+    {
+        return key($this->items);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function valid()
+    {
+        return !empty($this->items);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function rewind()
+    {
+        reset($this->items);
+    }
+
+    /**
+     * Get the next items (hydrated from doctrine repository)
+     *
+     * @param array       $esQuery
+     * @param string|null $searchAfter
+     *
+     * @return array
+     */
+    private function getNextItems(array $esQuery, $searchAfter = null)
+    {
+        $identifiers = $this->getNextIdentifiers($esQuery, $searchAfter);
+        if (empty($identifiers)) {
+            return [];
+        }
+
+        return $this->repository->loadItemsFromIdentifiers($identifiers);
+    }
+
+    /**
+     * Get the next identifiers from elasticsearch query
+     *
+     * @param array       $esQuery
+     * @param string|null $searchAfter
+     *
+     * @return array
+     */
+    private function getNextIdentifiers(array $esQuery, $searchAfter = null)
+    {
+        $esQuery['size'] = $this->pageSize;
+
+        if (!isset($esQuery['sort'])) {
+            $esQuery['sort'] = ['_uid' => 'asc'];
```

Comment:
we'll need to handle the sort at some point
but OK to not do here in your PR

we'll open another one